### PR TITLE
Feature/push/add echo client

### PR DESCRIPTION
**Pull request by TalhaAli00**

### CODE REVIEW

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/model/EchoResponse.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        @Json(name = "name")
+        val name: String
+    )
+
+    @JsonClass(generateAdapter = true)
+    data class Field(
+        @Json(name = "description")
+        val description: String,
+        @Json(name = "field")
+        val `field`: String,
```
Comment (jakobuid) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/model/EchoResponse.kt:28]: Just curious: why is there single quotation mark?

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/network/EchoService.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import retrofit2.http.DELETE
+import retrofit2.http.POST
+import retrofit2.http.Path
+
+interface EchoService {
+
+    @POST("893ad3af-9515-42d6-aaa2-06b53494b9a5/clients")
+    suspend fun register(@Body echoClientsBody: EchoBody): Response<EchoResponse>
+
+    @DELETE("893ad3af-9515-42d6-aaa2-06b53494b9a5/clients/{clientId}")
```
Comment (jakobuid) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/network/EchoService.kt:16]: This "893ad3af-9515-42d6-aaa2-06b53494b9a5" looks like some static value. What is that?

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import retrofit2.Retrofit
+
+internal object EchoClient: EchoInterface {
+    private val echoService by lazy { wcKoinApp.koin.get<Retrofit>().create(EchoService::class.java) }
+    private val clientId by lazy { requireNotNull(wcKoinApp.koin.get<SharedPreferences>().getString(EchoInterface.KEY_CLIENT_ID, null)) }
+    private const val SUCCESS_STATUS = "SUCCESS"
+
+    internal fun initialize() {}
```
Comment (jakobuid) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:21]: What's the purpose of this method? When EchoClient should be initialized? Should it be a part of the CoreClient?

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import com.walletconnect.android.internal.common.scope
+import com.walletconnect.android.internal.common.wcKoinApp
+import com.walletconnect.foundation.common.model.Topic
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import retrofit2.Retrofit
+
+internal object EchoClient: EchoInterface {
+    private val echoService by lazy { wcKoinApp.koin.get<Retrofit>().create(EchoService::class.java) }
+    private val clientId by lazy { requireNotNull(wcKoinApp.koin.get<SharedPreferences>().getString(EchoInterface.KEY_CLIENT_ID, null)) }
```
Comment (jakobuid) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:18]: Can't we create a echo koin module that'll be initialized inside the initialize function of EchoClient?

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                } else {
+                    onError(IllegalArgumentException(response.errorBody()?.string()))
+                }
+            }
+        } catch (e: Exception) {
+            onError(e)
+        }
+    }
+
+    override fun unregister(onSuccess: () -> Unit, onError: (Throwable) -> Unit) {
```
Comment (jakobuid) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:45]: When unregister should be called?

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                } else {
+                    onError(IllegalArgumentException(response.errorBody()?.string()))
+                }
+            }
+        } catch (e: Exception) {
+            onError(e)
+        }
+    }
+
+    override fun decryptMessage(topic: String, message: String, onSuccess: (String) -> Unit, onError: (Throwable) -> Unit) {
```
Comment (jakobuid) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:65]: Why decryptMessage method is needed here?  What's the purpose of it? There is already one in the PushWalletInterface.

File: push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletInterface.kt [hunk 7]
```diff
@@ -25,5 +25,7 @@ interface PushWalletInterface {
 
     fun delete(params: Push.Wallet.Params.Delete, onError: (Push.Model.Error) -> Unit)
 
-    fun decryptMessage(params: Push.Wallet.Params.DecryptMessage, onSuccess: (Push.Model.Message) -> Unit, onError: (Push.Model.Error) -> Unit)
+    fun registerFcmToken(params: Push.Wallet.Params.FcmToken, onSuccess: () -> Unit, onError: (Push.Model.Error) -> Unit)
```
Comment (jakobuid) [inline push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletInterface.kt:28]: Isn't it enough to just call a register method from EchoClient? Instead of just wrapping it in the PushWalletProtocol.

Also, from what I read in specs, register is not a part of the PushWalletClient.

File: push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletProtocol.kt [hunk 8]
```diff
@@ -36,8 +34,6 @@ class PushWalletProtocol : PushWalletInterface {
                 walletEngineModule(),
```
Comment (jakobuid) [inline push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletProtocol.kt:34]: // TODO: Commented out until we merge PR to handle multiple versions of dependnecy
//                pushCommonModule(),
//                cryptoModule(),

Let's clean this up since the PR responsible for consolidating androidCore-impl into androidCore-sdk is merged

File: push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletProtocol.kt [hunk 9]
```diff
@@ -95,12 +91,20 @@ class PushWalletProtocol : PushWalletInterface {
...
+
+        echoClient.register(params.fcmAccessToken, onSuccess) { error ->
+            onError(Push.Model.Error(error))
+        }
+    }
+
+    override fun decryptMessage(params: Push.Wallet.Params.DecryptMessage, onSuccess: (String) -> Unit, onError: (Push.Model.Error) -> Unit) {
+        val echoClient = wcKoinApp.koin.get<EchoInterface>()
+
+        echoClient.decryptMessage(params.topic, params.encryptedMessage, onSuccess) { error ->
```
Comment (jakobuid) [inline push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletProtocol.kt:105]: I don't see a need for decrypt method inside EchoClient

File: sign/wallet/build.gradle.kts [hunk 10]
```diff
@@ -44,13 +45,16 @@ dependencies {
     implementation(project(":sign:samples_common"))
 
     debugImplementation(project(":androidCore:sdk"))
-    debugImplementation(project(":sign:sdk"))
+    debugImplementation(project(":web3:wallet"))
```
Comment (jakobuid) [inline sign/wallet/build.gradle.kts:48]: Why this change?

Also, I think it would be great to have Push integration in our new Web3Wallet app, so that people could actually test the whole flow.

File: sign/wallet/src/main/kotlin/com/walletconnect/wallet/WalletFirebaseMessagingService.kt [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+import androidx.core.app.NotificationCompat
+import com.google.firebase.messaging.RemoteMessage
+import com.walletconnect.push.common.Push
+import com.walletconnect.push.wallet.client.PushMessageService
+import com.walletconnect.wallet.ui.host.WalletSampleActivity
+import kotlin.random.Random
+import kotlin.random.nextUInt
+
+@SuppressLint("MissingFirebaseInstanceTokenRefresh")
+class WalletFirebaseMessagingService: PushMessageService() {
```
Comment (jakobuid) [inline sign/wallet/src/main/kotlin/com/walletconnect/wallet/WalletFirebaseMessagingService.kt:21]: This idea of creating our own messaging service is very nice. I like it, good job.

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/di/CoreNetworkModule.kt [hunk 12]
```diff
@@ -86,4 +88,11 @@ fun coreAndroidNetworkModule(serverUrl: String, jwt: String, connectionType: Con
         ConnectivityState(androidApplication())
     }
 
+    single {
```
Comment (Elyniss) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/di/CoreNetworkModule.kt:91]: As there will be another Retrofit instance in Chat SDk for Keyserver. Could we use named dependency here?

File: build.gradle.kts [hunk 13]
```diff
@@ -16,6 +16,7 @@ buildscript {
         classpath("org.jetbrains.dokka:dokka-core:$dokkaVersion")      // TODO: Leave version until AGP 7.3 https://github.com/Kotlin/dokka/issues/2472#issuecomment-1143604232
         classpath("org.jetbrains.dokka:dokka-gradle-plugin:$dokkaVersion")
         classpath("com.squareup.sqldelight:gradle-plugin:$sqlDelightVersion")
+        classpath("com.google.gms:google-services:4.3.14")
```
Comment (Elyniss) [inline build.gradle.kts:19]: Let's extract version

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/model/EchoResponse.kt (hunk 1 above)
Comment (TalhaAli00) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/model/EchoResponse.kt:28]: since field is a Kotlin keyword, the single quotation marks are needed to have a variable of the same name

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/network/EchoService.kt (hunk 2 above)
Comment (TalhaAli00) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/network/EchoService.kt:16]: this is the FCM endpoint for echo server

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt (hunk 3 above)
Comment (TalhaAli00) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:21]: I kept initialize but it makes sense to remove it now

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt (hunk 5 above)
Comment (TalhaAli00) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:45]: unregister is for wallets to tell Echo server to not send anymore notifications

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt (hunk 6 above)
Comment (TalhaAli00) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:65]: That's a good callout. I've removed it from EchoClient

File: push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletProtocol.kt (hunk 8 above)
Comment (TalhaAli00) [reply push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletProtocol.kt:34]: removed

File: sign/wallet/build.gradle.kts (hunk 10 above)
Comment (TalhaAli00) [reply sign/wallet/build.gradle.kts:48]: missed this during rebase. Fixed

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/di/CoreNetworkModule.kt (hunk 12 above)
Comment (TalhaAli00) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/di/CoreNetworkModule.kt:91]: done

File: push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletProtocol.kt (hunk 9 above)
Comment (TalhaAli00) [reply push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletProtocol.kt:105]: moved logic to PushWallet client

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt (hunk 4 above)
Comment (TalhaAli00) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:18]: What do you mean by having an echo koin module? A koin module defines dependencies

File: push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletInterface.kt (hunk 7 above)
Comment (TalhaAli00) [reply push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushWalletInterface.kt:28]: You're right about not being in the specs. I've removed this function and reverted the service to use `CoreClient.Echo`

**Review by Elyniss — APPROVED** [review]

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt (hunk 4 above)
Comment (jakobuid) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:18]: It's not clear to me how developers initialize EchoClient. If it's by calling the init function we could get rid of lazy koin initialization and create a koin module for that. Such a module could be called inside the init function.

I want to avoid any potential race conditions on initialization and assure the ride order on components initialization.

Comment (jakobuid) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:18]: Ok, I've just noticed that the initialization of EchoClient happens in CoreClient. This makes sense, so could we add EchoService and clientId to Koin already inside CoreClient?
Does it make sense to you?

File: sign/wallet/src/main/kotlin/com/walletconnect/wallet/WalletFirebaseMessagingService.kt [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+    override fun registeringFailed(token: String, throwable: Throwable) {
+        Log.d(TAG, token)
+    }
+
+    override fun onDecryptedMessage(decryptedMessage: String, originalMessage: RemoteMessage) {
+        Log.d(TAG, decryptedMessage)
+    }
+
+    override fun onUnencryptedMessage(message: Push.Model.Message, originalMessage: RemoteMessage) {
```
Comment (jakobuid) [inline sign/wallet/src/main/kotlin/com/walletconnect/wallet/WalletFirebaseMessagingService.kt:40]: I think from the developer's perspective it shouldn't matter whether the message was decrypted by the SDK or was unencrypted when received.
It'd be more convenient to have one callback such as onMessageReceived that has only a Push.Model.Message as a parameter.
The output of decrypting the encrypted message should be the same as the unencrypted message received by the SDK.

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt (hunk 4 above)
Comment (TalhaAli00) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/echo/EchoClient.kt:18]: Makes sense. I've added a Echo module

File: sign/wallet/src/main/kotlin/com/walletconnect/wallet/WalletFirebaseMessagingService.kt (hunk 14 above)
Comment (TalhaAli00) [reply sign/wallet/src/main/kotlin/com/walletconnect/wallet/WalletFirebaseMessagingService.kt:40]: That's a valid point. I've consolidated the two functions and handle both cases now

File: push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushMessageService.kt [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+                onError(e, message)
+            }
+        }
+    }
+
+    abstract fun newToken(token: String)
+
+    abstract fun registeringFailed(token: String, throwable: Throwable)
+
+    abstract fun ondMessage(message: Push.Model.Message, originalMessage: RemoteMessage)
```
Comment (jakobuid) [inline push/sdk/src/main/kotlin/com/walletconnect/push/wallet/client/PushMessageService.kt:56]:
```suggestion
    abstract fun onMessage(message: Push.Model.Message, originalMessage: RemoteMessage)
```

**Review by jakobuid — APPROVED** [review]
I left one last comment about typo, approved 👍
